First phase of Hotel Booking module

mazurewa opened the pull request.

mazurewa committed:
Add files via upload

jacekmlynek commented on HotelBooking/HotelBooking.AcceptanceTests/App.config:
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <configSections>
+    <section name="specFlow" type="TechTalk.SpecFlow.Configuration.ConfigurationSectionHandler, TechTalk.SpecFlow" />
```
Specflow tests - Oh My Good Ewa bierzesz mnie pod wlos ;)

jacekmlynek replied:
Ok. To tyle na ta chwile :).

Nie jest to mega dokladne review ale dotyka chyba najwazniejszych problemow. 
Jesli cos jest nie jasne wal smialo

jacekmlynek commented on HotelBooking/HotelBooking.AcceptanceTests/Features/ProduceBookingResult.feature:
```diff
@@ -0,0 +1,25 @@
+﻿Feature: ProduceBookingResult
```
Pare ogolnych uwag co do testow akceptacyjnych:
* My piszemy testy akceptacyjne jako testy kopmponetu bardziej niz testy end to end. Jesli chcesz zglebic wiedze mozesz zaczac od tego linka https://martinfowler.com/articles/microservice-testing/
* pierwszym obowiazkiem takich testow nie jest "testowanie" (choc to oczywiscie tez) ale wyjasnienie komus kto bedzie korzystal z naszej apki albo webservice jak to dziala. Czyli jaki jest publiczy interfejs i kontrakt!
* podaba mi sie scenario "Failed reservation - email is in invalid format" (dotyczy wlasnie interfejsu publicznego), nie podoba mi sie krok "Failed reservation - hotel does not have all required operations included" (konfiguracja hotelu jest czymś na co client twojej apki nie ma wplywu. Dla niego blad jak kazdy inny)
* Chyba byl bym tez bardzie wylewny jesli chodzi o mozliwosci mojej aplikacji. Np. podzilbym features na:
  * BookingHotelRoomFeature
  * PaymentFerature
  * ConfirmationEmailFeatur 
* Same zas Scenarios bardziej bym celowal w cos ala:
  * Scenario: Booking with not authorized payment
  * Scenario: Booking with not delivered email
  * Scenario: Custome email in worng format

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationEqualityComparerTests.cs:
```diff
@@ -0,0 +1,58 @@
+            operationA.Stub(x => x.OperationName).Return("name");
+            IOperation operationB = MockRepository.GenerateStub<IOperation>();
+            operationB.Stub(x => x.OperationName).Return("name");
+            sut = new OperationEqualityComparer();
```
Pierdola ale aby nie powtarzac tego w kazdym testie mozesz umiescic to w hooku SetUp, ktory jest wykonywany przed kazdym testem

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationsManagerTests.cs:
```diff
@@ -0,0 +1,246 @@
+        }
+
+        [Test]
+        public void Should_GetOrderedOperationsForHotel_When_Called()
```
co to when called tutaj wnosi? Jest cos takiego jak WhenNotCalled? Mysle ze mozna z tego zrezygnowac

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationsManagerTests.cs:
```diff
@@ -0,0 +1,246 @@
+        {
+            var operationsToRun = new List<IOperation>();
+            operationsProvider.Stub(x => x.GetOrderedOperations(reservation.Hotel)).Return(operationsToRun);
+            operationsProvider.Stub(x => x.ContainsAllRequiredOperations(operationsToRun)).Return(false);
```
Mysle ze juz ten test to pierwszy sygnal ze cos nie dokonca jest ok z OperationProvider. W tej samej klasie robie twa stuby, przyczym wejsciem do 2 jest rezultat z pierwszego. Nie otwierajac nawet kodu juz zastanawiam sie czy klasa Providera nie moze zapewnic mi "oprawnych operacji". Otwarice kodu tylko powierdza. Moja sugestia wystawilbym tylko jedna metode i w niej zrobil cala robote, albo jesli to ma jakis wiekszy sens stworzyl dedykowana klase dla tego sprawdzenia (raczej watpie)

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationsManagerTests.cs:
```diff
@@ -0,0 +1,246 @@
+        [Test]
+        public void Should_ProcessThreeOperations_When_FourOperationsSpecifiedByHotelAndProcessAbortedOnThird()
+        {
+            var operation1 = MockRepository.GenerateMock<IOperation>();
```
Poprzedni test juz nie jest zabardzo czytelny a ten to juz zwlaszcza. Po pierwsze  uzywanie liczb w nazwach testow jest srednim pomyslem, o ile nie sa to wazne wartosci graniczne (wieksze rowne itp.) 3 chyba nie jest tu granica?

Wystarczajace bylo by pokazanie poprostu tego ze jesli  poprszednia operacja zwraca Abort to nie wykonuje sie nastepna. Czy jest ich 3 czy 4 nie ma to znaczenia!

Czyli ten test nazwalbym:

WhenOpewrationFailedWithAbort_ShouldStopProcessing()

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationsManagerTests.cs:
```diff
@@ -0,0 +1,246 @@
+
+        [Test]
+        public void Should_ReturnTwoOperationsResults_When_ProcessAbortedOnSecond()
+        {
```
:) Ahh taki test nie pomaga. I znowu czy to 2 czy 3 nie ma znaczenia. Chetnie wyjasnie face to face bo tu chyba mi sie nie uda.

jacekmlynek commented on HotelBooking/HotelBooking.App.Tests/OperationsProcessingTests/OperationsProviderTests.cs:
```diff
@@ -0,0 +1,150 @@
+        }
+
+        [Test]
+        public void Should_ReturnFourOperations_When_FourOperationsSpecifiedByHotelAndAllAvailable()
```
Ten test maskuje cos co dla niego jest istotne jak wnosze po nazwie - "AvailableHotels". Z jednej storny wyraznie definuje operacje w hotelu z drugiej zas gdzies tam za pomoca helpera ustawiam je. Wedlug mnie test powinien pokazywac ze tu i tu ustawiam ta sama kolekcje operacji.

jacekmlynek commented on HotelBooking/HotelBooking.App/DataObjects/BookingResult.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class BookingResult
+    {
+        public Result OverallResult { get; set; }
```
Zaproponowalbym zeby ten rezultat bym obliczny a nie ustawiany

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/EmailSender.cs:
```diff
@@ -0,0 +1,26 @@
+
+        public override bool IsRequiredToSucceed => false;
+        public override string OperationName => OperationCode.SendingEmail;
+        public override Func<Reservation, bool> Execute => SendConfirmationEmail;
```
ta metoda/property jest tylko wykorzystywana w klasie bazowej. Oprocz testow, ktore mozna napisac inaczej, nie ma sensu aby byla publiczna.

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/EmailSender.cs:
```diff
@@ -0,0 +1,26 @@
+            _emailSender = emailSender;
+        }
+
+        public override bool IsRequiredToSucceed => false;
```
Ta property jest uzywana tylko w klasie bazowej nie musi byc publiczna

jacekmlynek replied:
W zadaniu to bardziej konfiguracja, caly algorytm procesowania mial decydowac czy operacja jest required czy nie, niz sama operacja.  To takze sprawa koncepcyjna

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/EmailSender.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        private IExternalEmailSystem _emailSender;
+
+        public EmailSender(IExternalEmailSystem emailSender)
```
To dotyczy chyba wszystkich twochich operacji. Nie wiem jakie plan mialas w glowie jesli deleguje cala funksjonalnosc do wspolpraownika to poza IsRequired nie widze za co ta klasa jest odpowiedzialna. Innymi slowy dla mnie to chyba troche za duzo tutaj delegacji ale to sprawa mocno koncpcyjna - do dyskusji

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/ExternalSystems/IExternalEmailSystem.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace HotelBooking.App.Operations.ExternalSystems
+{
+    public interface IExternalEmailSystem
```
To jest wlasnie przyklad chyba nie dokonca dobrej nazwy w tej aplikacji. External nie powinien byc tu podmiotem. To jest tylko szczegol. Rownie dobrze moze byc Internal. Dlatego nazwalbym to poprostu IEmailSystem.

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/ExternalSystems/IExternalHotelPriceValidator.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace HotelBooking.App.Operations.ExternalSystems
+{
+    public interface IExternalHotelPriceValidator
```
To samo jak wyzej. Wszystkie interfejsy w nazwie External maja ta przypadlosc

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/OperationBase.cs:
```diff
@@ -0,0 +1,36 @@
+                OperationName = OperationName,
+                ShouldAbortBookingProcess = IsRequiredToSucceed && executionResult == Result.Failure
+            };           
+            bookingResult.OperationResults.Add(OperationResult);
```
Mysle ze pojedyncza operacja nie powinna wiedziec jak agregowac bledy. Powinna to robic klasa wyzej - OperationManager. chetnie to rozwine jak nie jest to jasne ale tak na szybko logika agregacji moze sie zmieniac z czasem co nie powinno miec wplywu na operacje. pozatym pozbedziesz modyfikowania bookingResult w tej metodzie, ktory dobrze by bylo jak by byl immutable.

jacekmlynek commented on HotelBooking/HotelBooking.App/Operations/OperationBase.cs:
```diff
@@ -0,0 +1,36 @@
+            WriteOperationResultToConsole();            
+        }
+
+        private void WriteOperationResultToConsole()
```
To cos co napewno warto zrobic. W wielu miejsach wstawiasz sobie tak lekka reka odwolanie do konsoli, tym samym uzalezniajac corowe klasy twojej apki od takiej infrastrukturalnej rzeczy jak consola. Co mozna zrobic.

W tym konkretnym przypadku zwrocilbym rezultat moze nawet do pliku programu i tam mialbym jakiegos formattera ktory by mi to sformatowal na cosole. W innych przypdkach, tam gdzie wykorzystuje ta informacje jak DEBUG INFO wstrzyknolbym jako zalerznosc cos w stylu ILogger albo IDebugInfoBuilder

jacekmlynek commented on HotelBooking/HotelBooking.App/OperationsProcessing/OperationsManager.cs:
```diff
@@ -0,0 +1,56 @@
+            var orderedOperations = _operationsProvider.GetOrderedOperations(reservation.Hotel);
+            var bookingResult = new BookingResult();
+
+            if (_operationsProvider.ContainsAllRequiredOperations(orderedOperations))
```
Zobczn moj komentarz a w klasie testow a propo tej metody

jacekmlynek commented on HotelBooking/HotelBooking.App/OperationsProcessing/OperationsManager.cs:
```diff
@@ -0,0 +1,56 @@
+                {
+                    operation.Process(reservation, bookingResult);
+
+                    if (operation.OperationResult.ShouldAbortBookingProcess)
```
to jest 3 zagnizdzenie w tej metodzie co moze czynnic ja delikatnie mniej czytelna. Moja rada wychodz z metod tak szybko jak to mozliwe.

``` java
 var orderedOperations = _operationsProvider.GetOrderedOperations(reservation.Hotel);

            if (!_operationsProvider.ContainsAllRequiredOperations(orderedOperations))
            {
                return new BookingResult();
            }

            var bookingResult = new BookingResult();
            foreach (var operation in orderedOperations)
            {
                operation.Process(reservation, bookingResult);

                if (operation.OperationResult.ShouldAbortBookingProcess)
                {
                    AbortProcess(reservation, operation.OperationName);
                    break;
                }
            }

            DeduceOverallBookingResult(bookingResult);

            return bookingResult;
```

jacekmlynek reviewed (commented):
Troche dalem ciala i zaczolem dwa review. Przeklej ci wstep z pierwszego abys miala pod jednym.

Ok Troche mi to zajelo ale juz mam.

Pare takich ogolnych mysli:
* jestem pod wrazeniem:
  * widac ze bardzo sie napracowalas - jest to chyba pierwsze rozwiazanie tego zadania, przy ktorym musialem posiedziec wiecej niz godzine.
  * staralas sie byc konsekwetna w calym projekcie  - konsekwencja spójny pomysł na applikacje/zadanie
  * wydaje mi się jak dobrze pamiętam że nie piszesz testow na codzien, a tu proszę:
    * sut 
    * jak na pierwszy raz calkiem dobre nazwy tesow
    *  mock/stubs
* pare rzeczy, które rzucają sie na początku w oczy:
  * było by fajnie jak tworzac appke dodała byś do README pare zdan  - co wedlug ciebie jest jej celem. W tym przypadku dla mnie pomocne bylo by, jako ze nie rozmawialismy na ten temat, przeczytac pare twoich ogolnych zalozen.
  * projeckt jest calkiem niezle podzielony ale:
    * Jesli Projeckt Domain ma cos wspolnego z DDD albo po prostu  jest "srodkiem/corem" dla twojego projektu to do niego radzilm bym przeniesc wiekszosc rzeczy z katalogow "Reservation" i "Operations", bo to jest istota tej aplikacji a nie w Modelu ktory jest durnym DTO czy w warstwie dostepu do danych.
    * Nie bardzo pasuje mi umiejscowienie "ExternalSystem" obok czegos takiego jak PriceValdiator albo EmailSender. Te ze cos jest wykonywane przez zewnetrzny system to taki szczegół (infrastrukturalna rzecz bardziej) natomiast sam fakt tego (valdiacja ceny, wyslanie maila) jest jak najbardziej opisem domenowym/corowym. Czuje że cię stracilem przy tej uwadze. Spoko  mogę to wyjaśnic face to face :)
    * Nie ma sensu robic katalogu takiego jak Enums to nie ma nic wspolnego z HotelBooking, ten enum mozna umiesic juz w jednym z istnijacych katalogow. 
   * Interfejsy - Ahh kocham interfejsy ale nie ma co nimi srać na lewo i prawo bo sie tylko wprowadza kogoś w błąd. Jesli klasy pochodza z tej smej warstwy i maja tylko jedna realizacje - to nie ma co komus sciemniac ze bazujesz na interfejsie. Co to za interfejs jesli masz w nim dokladnie to co masz w jednej jedynej klasie, malo tego w pelni dostepnej? Spoko wiem ze to nie latwo przelknac Jakub K. tez potrzebowal czasu, ale chyba sie z tym wpelni zgadza teraz :)

To tyle slowem wstepu, naprawde dobry kawałek kodu, dalsze komentarze beda juz do konkretnego lini lub klasy.

mazurewa committed:
Adding details to Readme, removing interfaces of External operations

mazurewa committed:
Validating required operation in OperationProvider

Tests not updated

mazurewa committed:
Moving calculation of overall boking result to BookingResult.OverallResult getter

mazurewa committed:
Removal of External prefix, renaming catalog to Services

mazurewa committed:
Moving opeartion results aggregation logic to OperationsManager

mazurewa committed:
Code review follow ups

Moving core functionality to Domain project, App as console application, removing services and OperationBase, adding logger

mazurewa committed:
Acceptance tests updates

Split into separate features

mazurewa committed:
Readme updated
